diff --git a/core/lib/Drupal/Core/Recipe/RecipeRunner.php b/core/lib/Drupal/Core/Recipe/RecipeRunner.php
index 5670c1f871..c5bf0d6322 100644
--- a/core/lib/Drupal/Core/Recipe/RecipeRunner.php
+++ b/core/lib/Drupal/Core/Recipe/RecipeRunner.php
@@ -2,8 +2,10 @@
 
 namespace Drupal\Core\Recipe;
 
+use Drupal\Core\Config\ConfigImporter;
 use Drupal\Core\Config\FileStorage;
 use Drupal\Core\Config\InstallStorage;
+use Drupal\Core\Config\StorageComparer;
 use Drupal\Core\Config\StorageInterface;
 
 /**
@@ -23,9 +25,37 @@ final class RecipeRunner {
    *   The recipe to apply.
    */
   public static function processRecipe(Recipe $recipe): void {
-    static::processRecipes($recipe->recipes);
-    static::processInstall($recipe->install, $recipe->config->getConfigStorage());
-    static::processConfiguration($recipe->config);
+    /** @var \Drupal\Core\Config\Checkpoint\CheckpointStorageInterface $checkpoint_storage */
+    $checkpoint_storage = \Drupal::service('config.storage.checkpoint');
+    $checkpoint = $checkpoint_storage->checkpoint('Backup');
+
+    try {
+      static::processRecipes($recipe->recipes);
+      static::processInstall($recipe->install, $recipe->config->getConfigStorage());
+      static::processConfiguration($recipe->config);
+    }
+    catch (InvalidConfigException $e) {
+      $checkpoint_storage->setReadFromCheckpoint($checkpoint);
+
+      $comparer = new StorageComparer($checkpoint_storage, \Drupal::service('config.storage'));
+      $comparer->reset();
+
+      $config_importer = new ConfigImporter(
+        $comparer,
+        \Drupal::service('event_dispatcher'),
+        \Drupal::service('config.manager'),
+        \Drupal::lock(),
+        \Drupal::service('config.typed'),
+        \Drupal::moduleHandler(),
+        \Drupal::service('module_installer'),
+        \Drupal::service('theme_handler'),
+        \Drupal::translation(),
+        \Drupal::service('extension.list.module'),
+        \Drupal::service('extension.list.theme'),
+      );
+      $config_importer->import();
+      $e->rethrow();
+    }
     static::processContent($recipe->content);
   }
 
