diff --git a/src/Controller/InstallerController.php b/src/Controller/InstallerController.php
index aa32429..066a7f7 100644
--- a/src/Controller/InstallerController.php
+++ b/src/Controller/InstallerController.php
@@ -307,14 +307,6 @@ class InstallerController extends ControllerBase {
    *   Status message.
    */
   public function begin(string $composer_namespace, string $project_id): JsonResponse {
-    // @todo Expand to support other plugins in https://drupal.org/i/3312354.
-    $source = $this->enabledSourceHandler->getCurrentSources()['drupalorg_mockapi'] ?? NULL;
-    if ($source === NULL) {
-      return new JsonResponse(['message' => "Cannot download $project_id from any available source"], 500);
-    }
-    if (!$source->isProjectSafe($project_id)) {
-      return new JsonResponse(['message' => "$project_id is not safe to add because its security coverage has been revoked"], 500);
-    }
     $stage_available = $this->installer->isAvailable();
     if (!$stage_available) {
       $requiring_metadata = $this->projectBrowserTempStore->getMetadata('requiring');
